homeassistant/core.py

```diff
@@ -1138,14 +1162,19 @@ def async_set(
         new_state = str(new_state)
         attributes = attributes or {}
         old_state = self._states.get(entity_id)
-        if old_state is None:
+        if old_state is None or ATTR_RESERVED in old_state.attributes:
+            old_state = None
             same_state = False
             same_attr = False
             last_changed = None
         else:
             same_state = old_state.state == new_state and not force_update
             same_attr = old_state.attributes == MappingProxyType(attributes)
             last_changed = old_state.last_changed if same_state else None
+            if ATTR_RESERVED in attributes:
```

Comment:
Rather than reserving an attribute, I could add `.new` or `.reserved` property to the `State` object.  The downside would be that every State would have the attribute and its only used during state creation.

Reply:
Why is `ATTR_RESERVED` needed, can't we just let `async_reserve` set state to `None`?

Reply:
I think that we can add setting it to `None`, and then add a new function to the state machine that checks if an entity ID is taken.

Reply:
I'll try it with None. I was a bit worried about that since callers may always expect a string

Reply:
I meant as the actual state object in the state machine. That way if you ask for the state, you still get `None` as that's in the dict. Although now we need to filter that in iterations 🤔

Reply:
Maybe we need a reservations dict

Reply:
We could make one at the entity component level and pass it into platforms. Doesn't need to be in core perse.

Reply:
I thought about it, and I think the state machine needs to be the source of truth here so I added it there.

Reply:
There are two places that generate/check for entity ids that also needed to be updated well.